### PR TITLE
Api for leave controller 

### PR DESCRIPTION
**Pull request by ror-uday**

### CODE REVIEW

File: app/controllers/api/v1/leafs_controller.rb [hunk 1]
```diff
@@ -0,0 +1,58 @@
+# frozen_string_literal: true
+
+module Api
+  module V1
+    class LeafsController < ApplicationController
+      # before_action :authenticate_employee!
+      skip_before_action :verify_authenticity_token
+
+      def index
+        current_employee
```
Comment (rorshaan) [inline app/controllers/api/v1/leafs_controller.rb:10]: Do we need this line?

File: app/controllers/api/v1/leafs_controller.rb [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        data = Leaf.where(employee_id: current_employee.id).order('created_at DESC')
+        render json: {
+          data: serializer_data(data, serializer),
+          message: ['leaf list '], status: 200, type: 'Success'
+        }
+      end
+
+      def create
+        data = Leaf.new(leaf_params)
```
Comment (rorshaan) [inline app/controllers/api/v1/leafs_controller.rb:19]: You you have used a wrong variable? 'data'
please use proper naming convention

File: app/controllers/api/v1/leafs_controller.rb [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+            data: serializer_data(data, serializer),
+            message: ['Successfully leave applied '], status: 200, type: 'Success'
+          }
+        else
+          render json: { message: ['leave not applied'] }, status: 404, type: 'failed'
+        end
+      end
+
+      def update
+        data = Leaf.find(params[:id])
```
Comment (rorshaan) [inline app/controllers/api/v1/leafs_controller.rb:31]: naming convention

File: app/serializers/api/v1/leaf_serializer.rb [hunk 4]
```diff
@@ -0,0 +1,10 @@
+# frozen_string_literal: true
+
+module Api
+  module V1
```
Comment (rorshaan) [inline app/serializers/api/v1/leaf_serializer.rb:4]: Have you run the robocop?

File: spec/v1/leafs_spec.rb [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+  end
+
+  before do
+    sign_in(employee)
+  end
+
+  describe '#index' do
+    it 'show all leave ' do
+      get :index
+      expect(response.status).to eq(200)
```
Comment (rorshaan) [inline spec/v1/leafs_spec.rb:21]: Why you are not expecting with objects?

File: spec/v1/leafs_spec.rb [hunk 6]
```diff
@@ -0,0 +1,54 @@
+# frozen_string_literal: true
+
```
Comment (rorshaan) [inline spec/v1/leafs_spec.rb:2]: Expect with objects as well in every spec.

**Review by rorshaan — CHANGES_REQUESTED** [review]

File: app/controllers/api/v1/leafs_controller.rb (hunk 1 above)
Comment (ror-uday) [reply app/controllers/api/v1/leafs_controller.rb]: changes done

File: app/controllers/api/v1/leafs_controller.rb (hunk 2 above)
Comment (ror-uday) [reply app/controllers/api/v1/leafs_controller.rb]: done sir

File: spec/v1/leafs_spec.rb (hunk 6 above)
Comment (ror-uday) [reply spec/v1/leafs_spec.rb:2]: changes done

File: app/serializers/api/v1/leaf_serializer.rb (hunk 4 above)
Comment (ror-uday) [reply app/serializers/api/v1/leaf_serializer.rb:4]: yes sir

File: app/controllers/api/v1/leafs_controller.rb [hunk 7]
```diff
@@ -0,0 +1,57 @@
+# frozen_string_literal: true
+
+module Api
+  module V1
+    class LeafsController < ApplicationController
+      # before_action :authenticate_employee!
+      skip_before_action :verify_authenticity_token
+
+      def index
+        leaf_data = Leaf.where(employee_id: current_employee.id).order('created_at DESC')
```
Comment (ror-uday) [inline app/controllers/api/v1/leafs_controller.rb:10]: current_employee.leafs

File: app/controllers/api/v1/leafs_controller.rb [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+          if leaf_data.update(leave_status: 'cancelled')
+            render json: {
+              data: serializer_data(leaf_data, serializer),
+              message: ['leaf cancelled'], status: 200, type: 'Success'
+            }
+          else
+            render json: leaf_data
+          end
+        else
+          render json: { message: ['leave status not update'] }, status: 304, type: 'failed'
```
Comment (ror-uday) [inline app/controllers/api/v1/leafs_controller.rb:41]: why 304 used

Comment (ror-uday) [inline app/controllers/api/v1/leafs_controller.rb:41]: why 304 used

File: app/controllers/api/v1/leafs_controller.rb [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+      def update
+        leaf_data = Leaf.find(params[:id])
+        if leaf_data.from_date > Time.now
+          if leaf_data.update(leave_status: 'cancelled')
+            render json: {
+              data: serializer_data(leaf_data, serializer),
+              message: ['leaf cancelled'], status: 200, type: 'Success'
+            }
+          else
+            render json: leaf_data
```
Comment (ror-uday) [inline app/controllers/api/v1/leafs_controller.rb:38]: Add error code

Comment (ror-uday) [inline app/controllers/api/v1/leafs_controller.rb:38]: Add error code

File: app/controllers/api/v1/leafs_controller.rb [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+            message: ['Successfully leave applied '], status: 200, type: 'Success'
+          }
+        else
+          render json: { message: ['leave not applied'] }, status: 404, type: 'failed'
+        end
+      end
+
+      def update
+        leaf_data = Leaf.find(params[:id])
+        if leaf_data.from_date > Time.now
```
Comment (ror-uday) [inline app/controllers/api/v1/leafs_controller.rb:31]: move in model

File: app/serializers/api/v1/leaf_serializer.rb [hunk 11]
```diff
@@ -0,0 +1,10 @@
...
+    class LeafSerializer < ActiveModel::Serializer
+      attributes :id, :leave_type, :from_date, :till_date, :leave_starts, :leave_end, :total_days, :resion, :leave_status,
```
Comment (ror-uday) [inline app/serializers/api/v1/leaf_serializer.rb:6]: change type of column name

Comment (ror-uday) [inline app/serializers/api/v1/leaf_serializer.rb:6]: change type of column name

File: config/routes.rb [hunk 12]
```diff
@@ -33,6 +33,13 @@
     get '/emp-attendance/:id', to: 'attendences#show_attendence', as: 'show_attendance'
     get '/search', to: 'attendences#search'
   end
+
+  namespace :api do
+    namespace :v1 do
+      resources :leafs
```
Comment (ror-uday) [inline config/routes.rb:39]: use only

File: app/controllers/api/v1/leafs_controller.rb (hunk 10 above)
Comment (ror-uday) [reply app/controllers/api/v1/leafs_controller.rb:31]: changes done

File: app/serializers/api/v1/leaf_serializer.rb (hunk 11 above)
Comment (ror-uday) [reply app/serializers/api/v1/leaf_serializer.rb]: changes done

File: app/controllers/api/v1/leafs_controller.rb (hunk 10 above)
[the remark this answers is not in the document]
Comment (ror-uday) [reply app/controllers/api/v1/leafs_controller.rb]: done

**Review by rorpyll — APPROVED** [review]
MSD

**Pull request closed without merging**